### DISCUSSION TITLE
What is the expected indentation for chained method calls?

### DISCUSSION THREAD

**Discussion by Vyom-Yadav** [checkstyle/checkstyle#11320] [General]
Just a general discussion, consider a few examples-
Example-1
```java
obj.asd( // indent:x exp:x
).asda( // indent:x exp:x
); // indent:x exp:x      
```

Example-2 (4 represents `lineWrappingIndentation`)
```java
asd().asd().asda().asd( // indent:x exp:x
    ) // indent:x+4 exp:x+4              
    .asd(); // indent:x+4 exp:x+4       
```

Example-3
```java
obj.asd( // indent:x exp:x           
).asda( // indent:x exp:x    
    ) // indent:x+4 exp:x+4          
    .asd().asd( // indent:x+4 exp:x+4
    ); // indent:x+4 exp:x+4         
```
Example-4
```java
obj.foo() // indent:x exp:x
    .hoo( // indent:x+4 exp:x+4
    ).asd( // indent:x+4 exp:x+4
    ).fef(); // indent:x+4 exp:x+4
```

The main confusion is when `(` and `)` are on different lines, is there any style guide that checkstyle follows for indentation which mentions such cases, I wasn't able to find much information in https://google.github.io/styleguide/javaguide.html#s4.1-braces, https://www.oracle.com/technetwork/java/codeconventions-150003.pdf
Also please share your personal opinion on this, let's finalize and implement this.
@romani @strkkk  @nmancus1 @rnveach @pbludov

**Comment by rnveach** [comment]
Whatever looks nice.

Many people have different ideas on what the perfect indention is. We just need to accept anything that looks acceptable and flag the ones that are completely bad. Sometimes we look at what IDEs provide as the default formatting.

Since we are talking about implementation, we should be looking at the existing issues reported and see what has been provided. Also look at what all the IDEs do.

**Comment by Vyom-Yadav** [comment]
I tried a few things, I am sharing what I propose as the final indentation-
```java
obj.asd( // indent:x exp:x
    ).asd( // indent:x+4 exp:x+4
    ).asd( // indent:x+4 exp:x+4
        arg1 // indent:x+8 exp:x+8
    ); // indent:x+4 exp:x+4
```
```java
obj.asd() // indent:x exp:x
    .asd() // indent:x+4 exp:x+4
    .asd(arg1, arg2, // indent:x+4 exp:x+4
        arg3 // indent:x+8 exp:x+8
    ) // indent:x+4 exp:x+4
    .asd( // indent:x+4 exp:x+4
    ); // indent:x+4 exp:x+4
```
```java
obj.asd( // indent:x exp:x
); // indent:x exp:x
```
I am completely open to suggestions.

**Comment by romani** [comment]
I would consider all that cases as bad (my style).
I would try to not focus on indentation to avoid conflicting with Indentation Check and try to focus on when to ask ".mymethod" should start a line.

as used define location of `)` we just put next chained method that that indenation level and start new method on new line.
Examples from above updated to what I (subjective) would expect. 
```
obj.asd(
)
.asda(
);      
```
```
asd().asd().asda().asd( // indent:x exp:x
    ) // indent:x+4 exp:x+4              
    .asd(); // indent:x+4 exp:x+4       
```

```
obj.asd( // indent:x exp:x           
).asda( // indent:x exp:x    
    ) // indent:x+4 exp:x+4          
    .asd().asd( // indent:x+4 exp:x+4
    ); // indent:x+4 exp:x+4         
```

```
obj.foo() // indent:x exp:x
    .hoo( // indent:x+4 exp:x+4
    )
    .asd( // indent:x+4 exp:x+4
    )
    .fef(); // indent:x+4 exp:x+4
```

I would focus less on indentation (see simple rule above that might be useful, but better to not think on this to avoid overlaps/conflicts) and focus more on location of chained methods

Should be good to my mind (indentation excluded intentionally, to focus on what matter the most):
```
obj.method1().method2()
    .methods3()
    .methods4();
```

```
obj.method1().method2()
    .methods3()
    .methods4().methods5();
```

as you can see first and last line can be chaned metods on same line and only user knows where to wrap, but mostlikely we can presume that all methods in between is good for being on new line.
This is just one variant of style that I noticed. We can make Check that try to enforce exactly this style.

**Comment by Vyom-Yadav** [comment]
> I would try to not focus on indentation to avoid conflicting with Indentation Check and try to focus on when to ask ".mymethod" should start a line.

@romani Um this is not related to `ChainedMethodCallWrapCheck`, it's related to `IndentationCheck` only, `MethodCallHandler` in indentation check does not work properly (numerous issues in the issue tracker) so I raised this question so I could rewrite it to work properly. I like your method of indentation expect-
```java
obj.asd( // indent:x exp:x           
).asda( // indent:x exp:x    
    ) // indent:x+4 exp:x+4          
    .asd().asd( // indent:x+4 exp:x+4
    ); // indent:x+4 exp:x+4         
```
The second line's indentation is peculiar IMO, shouldn't it be `x+4`? Intellij formats this as-
```java
obj.asd( // indent:x exp:x           
    ).asda( // indent:x+4 exp:x+4    
    ) // indent:x+4 exp:x+4          
    .asd().asd( // indent:x+4 exp:x+4
    ); // indent:x+4 exp:x+4         
```
I have made good progress in `MethodCallHandler` at this point, am configuring `NewHandler` to synergize with `MethodCallHandler`, will try to finalize and create a PR as soon as we conclude this discussion.

**Reply by romani** [reply]
As soon as you put \n in middle of statement it is line wrap, so +4

```
obj.asd( // indent:x exp:x           
).asda( // indent:x exp:x
```

Should be:
```
obj.asd( // indent:x exp:x           
).asda( // indent:x exp:x+4
```
